File: src/main/java/pl/training/module6/factory/after/DeploymentProbeFactory.java
package pl.training.module6.factory.after;

import java.util.Objects;

public final class DeploymentProbeFactory {
    public DeploymentProbe create(ProbeKind kind, String target) {
        Objects.requireNonNull(kind, "kind");
        if (target == null || target.isBlank()) {
            String field = kind == ProbeKind.HTTP ? "endpoint" : "queueName";
            throw new IllegalArgumentException(field + " must not be blank");
        }

        return switch (kind) {
            case HTTP -> new HttpProbe(target);
            case QUEUE -> new QueueProbe(target);
        };
    }

    public enum ProbeKind {
        HTTP,
        QUEUE
    }

    private record HttpProbe(String endpoint) implements DeploymentProbe {
        @Override
        public String check() {
            return "http-ok:" + endpoint;
        }
    }

    private record QueueProbe(String queueName) implements DeploymentProbe {
        @Override
        public String check() {
            return "queue-ok:" + queueName;
        }
    }
}

File: csharp/src/Training.Module6/Factory/After/DeploymentProbeFactory.cs
namespace Training.Module6.Factory.After;

public sealed class DeploymentProbeFactory
{
    public IDeploymentProbe Create(ProbeKind kind, string? target)
    {
        if (string.IsNullOrWhiteSpace(target))
        {
            var field = kind == ProbeKind.Http ? "endpoint" : "queueName";
            throw new ArgumentException(field + " must not be blank");
        }

        return kind switch
        {
            ProbeKind.Http => new HttpProbe(target),
            ProbeKind.Queue => new QueueProbe(target),
            _ => throw new ArgumentOutOfRangeException(nameof(kind), kind, null)
        };
    }

    public enum ProbeKind
    {
        Http,
        Queue
    }

    private sealed record HttpProbe(string Endpoint) : IDeploymentProbe
    {
        public string Check() => "http-ok:" + Endpoint;
    }

    private sealed record QueueProbe(string QueueName) : IDeploymentProbe
    {
        public string Check() => "queue-ok:" + QueueName;
    }
}
